src/jit/copyprop.cpp

```diff
@@ -290,6 +290,147 @@ bool Compiler::optIsSsaLocal(GenTree* tree)
     return tree->IsLocal() && lvaInSsa(tree->AsLclVarCommon()->GetLclNum());
 }
 
+/**************************************************************************************
+ *
+ * Helper to get the LclVar for the Expr.
+ */
+bool Compiler::optTryGetCopyPropLclVars(GenTree* tree, GenTreeLclVarCommon*& srcLclVar, GenTreeLclVarCommon*& dstLclVar)
+{
+    if (!optTryGetCopyPropLclVar(tree->gtGetOp2(), srcLclVar))
+    {
+        return false;
+    }
+
+    if (!optTryGetCopyPropLclVar(tree->gtGetOp1(), dstLclVar))
+    {
+        return false;
+    }
+
+    return true;
+}
+
+/**************************************************************************************
+ *
+ * Helper to get the LclVar for the Oper.
+ */
+bool Compiler::optTryGetCopyPropLclVar(GenTree* op, GenTreeLclVarCommon*& lclVar)
+{
+    if (varTypeIsSIMD(op) || op->IsPhiNode())
+    {
+        // Leave SIMD and PhiNodes alone
+        return false;
+    }
+
+    if (op->OperIsBlk() && (op->AsBlk()->Addr()->OperGet() == GT_ADDR))
+    {
+        op = op->AsBlk()->Addr()->gtGetOp1();
+    }
+    if (op->OperIsIndir() && (op->AsIndir()->Addr()->OperGet() == GT_ADDR))
+    {
+        op = op->AsIndir()->Addr()->gtGetOp1();
+    }
+
+    if ((op->OperGet() == GT_LCL_VAR || op->OperGet() == GT_LCL_FLD))
```

Comment:
I don't see anything in this code that would make `GT_LCL_FLD` a valid optimization target, except in one particular case:
```
x0 = x1;
x2 = x0.f0;
```
that can usually be transformed into
```
x0 = x1;
x2 = x1.f0;
```
with some precautions in case `x0` and `x1` do not have the same type. I don't see what precautions have you taken to ensure that this and only this case is handled.

Reply:
Is this for the `optCopyPropThroughCopyBlk` version rather than `optCopyPropFoldCopyBlks`?

Which precautions would I need? 

It currently checks the are the exact same size; and that both are last use so converts
```
(def) x0 = x1;    (last use)
      x2 = x0.f0; (last use)
```
to 
```
      x2 = x1.f0; (last use)
```

Reply:
> Which precautions would I need?

For example, the following transform would be obviously incorrect (unless f0 happens to be the only field):
```
x0.f0 = x1.f0;
x2 = x0;
```
to
```
x0.f0 = x1.f0;
x2 = x1;
```

Reply:
So this is one of the LCL_FLD's it changes; removing `V446` in its entirely
```
CopyBlk based forward copy assertion for forward copy assertion for [006950] V446 @000003B1 by [006911] V444 @00000003.

***** (before)
N003 (  7,  5) [006957] -A------R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [006955] D------N----              +--*  LCL_VAR   struct V446 tmp426      d:1
N001 (  3,  2) [006911] -------N----              \--*  LCL_VAR   struct V444 tmp424      u:1 (last use) $6c8

N003 (  3,  4) [006966] -A--G---R---              *  ASG       byref  $3b1
N002 (  1,  1) [006965] D------N----              +--*  LCL_VAR   byref  V447 tmp427      d:1 $3b1
N001 (  3,  4) [006950] ------------              \--*  LCL_FLD   byref  V446 tmp426      u:1[+0] Fseq[_pointer, _value] (last use) $3b1

Copy propagated to:

***** (after)
N001 (  7,  5) [006957] ------------              *  NOP       void  

     (  3,  4) [006967] ------------              *  STMT      void  (IL   ???...  ???)
N003 (  3,  3) [006966] -A--G---R---              \--*  ASG       byref  $3b1
N002 (  1,  1) [006965] D------N----                 +--*  LCL_VAR   byref  V447 tmp427      d:1 $3b1
N001 (  3,  2) [008934] -------N----                 \--*  LCL_VAR   byref  V444 tmp424       (last use)
```

Reply:
So this is one of the LCL_FLDs it changes; removing `V446` in its entirely
```
CopyBlk based forward copy assertion for forward copy assertion for [006950] V446 @000003B1 by [006911] V444 @00000003.

***** (before)
N003 (  7,  5) [006957] -A------R---  *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [006955] D------N----  +--*  LCL_VAR   struct V446 tmp426      d:1
N001 (  3,  2) [006911] -------N----  \--*  LCL_VAR   struct V444 tmp424      u:1 (last use) $6c8

N003 (  3,  4) [006966] -A--G---R---  *  ASG       byref  $3b1
N002 (  1,  1) [006965] D------N----  +--*  LCL_VAR   byref  V447 tmp427      d:1 $3b1
N001 (  3,  4) [006950] ------------  \--*  LCL_FLD   byref  V446 tmp426      u:1[+0] Fseq[_pointer, _value] (last use) $3b1

Copy propagated to:

***** (after)
N001 (  7,  5) [006957] ------------  *  NOP       void  

     (  3,  4) [006967] ------------  *  STMT      void  (IL   ???...  ???)
N003 (  3,  3) [006966] -A--G---R---  \--*  ASG       byref  $3b1
N002 (  1,  1) [006965] D------N----     +--*  LCL_VAR   byref  V447 tmp427      d:1 $3b1
N001 (  3,  2) [008934] -------N----     \--*  LCL_VAR   byref  V444 tmp424       (last use)
```

Which is why I'm using creation and death; if its created in one statement for a copy and then dies in the next, it removes it and uses the original

Reply:
For `<ReadAsyncInternal>d__66:MoveNext` to does the following conversions
```
***** (before)
N003 (  7,  5) [002491] -A--G---R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [002490] D---G--N----              +--*  LCL_VAR   struct(AX) V16 loc15        
N001 (  3,  2) [002488] ------------              \--*  LCL_VAR   struct V112 tmp93       u:2 (last use) $456

N007 ( 13, 15) [002503] -A--G---R---              *  ASG       struct (copy) $VN.Void
N006 (  6,  7) [002502] n-----------              +--*  BLK(16)   struct
N005 (  3,  5) [002501] ------------              |  \--*  ADDR      byref  $8bd
N004 (  3,  4) [002494] D------N----              |     \--*  LCL_FLD   struct V112 tmp93       d:2[+0] Fseq[_value]
N003 (  6,  7) [002499] n---G-------              \--*  IND       struct $5c5
N002 (  3,  5) [002495] ----G-------                 \--*  ADDR      byref  $18f
N001 (  3,  4) [002498] -------N----                    \--*  LCL_FLD   struct V09 loc8         u:3[+0] Fseq[_value] (last use) $5c5

Copy propagated to:

***** (after)
N001 (  7,  5) [002491] ------------              *  NOP       void  

     ( 13, 15) [002504] ------------              *  STMT      void  (IL 0x2CE...  ???)
N005 ( 10, 10) [002503] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N004 (  3,  2) [005873] D---G--N----                 +--*  LCL_VAR   struct(AX) V16 loc15        
N003 (  6,  7) [002499] n---G-------                 \--*  IND       struct $5c5
N002 (  3,  5) [002495] ----G-------                    \--*  ADDR      byref  $18f
N001 (  3,  4) [002498] -------N----                       \--*  LCL_FLD   struct V09 loc8         u:3[+0] Fseq[_value] (last use) $5c5

CopyBlk based forward copy assertion for reverse copy assertion for [002426] V09 @00000003 by [002498] V09 @00000003.

***** (before)
N005 ( 10, 10) [002503] -A--G---R---              *  ASG       struct (copy) $VN.Void
N004 (  3,  2) [005873] D---G--N----              +--*  LCL_VAR   struct(AX) V16 loc15        
N003 (  6,  7) [002499] n---G-------              \--*  IND       struct $5c5
N002 (  3,  5) [002495] ----G-------                 \--*  ADDR      byref  $18f
N001 (  3,  4) [002498] -------N----                    \--*  LCL_FLD   struct V09 loc8         u:3[+0] Fseq[_value] (last use) $5c5

N003 (  7,  5) [002427] -A------R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [002426] D------N----              +--*  LCL_VAR   struct V09 loc8         d:3
N001 (  3,  2) [002424] ------------              \--*  LCL_VAR   struct V108 tmp89       u:2 (last use) $588

Copy propagated to:

***** (after)
N001 ( 10, 10) [002503] ------------              *  NOP       void  

     (  7,  5) [000268] ------------              *  STMT      void  (IL   ???...  ???)
N003 (  7,  5) [002427] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005874] D---G--N----                 +--*  LCL_VAR   struct(AX) V16 loc15        
N001 (  3,  2) [002424] ------------                 \--*  LCL_VAR   struct V108 tmp89       u:2 (last use) $588

CopyBlk based forward copy assertion for reverse copy assertion for [002465] V108 @00000003 by [002424] V108 @00000003.

***** (before)
N003 (  7,  5) [002427] -A--G---R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005874] D---G--N----              +--*  LCL_VAR   struct(AX) V16 loc15        
N001 (  3,  2) [002424] ------------              \--*  LCL_VAR   struct V108 tmp89       u:2 (last use) $588

N005 ( 10, 10) [002472] -A--G---R---              *  ASG       struct (copy) $VN.Void
N004 (  6,  7) [002471] n-----------              +--*  BLK(16)   struct
N003 (  3,  5) [002470] ------------              |  \--*  ADDR      byref  $8bc
N002 (  3,  4) [002465] D------N----              |     \--*  LCL_FLD   struct V108 tmp89       d:2[+0] Fseq[_value]
N001 (  3,  2) [002467] ----G--N----              \--*  LCL_VAR   struct(AX) V107 tmp88        $455

Copy propagated to:

***** (after)
N001 (  7,  5) [002427] ------------              *  NOP       void  

     ( 10, 10) [002473] ------------              *  STMT      void  (IL 0x2C4...  ???)
N003 (  7,  5) [002472] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005875] D---G--N----                 +--*  LCL_VAR   struct(AX) V16 loc15        
N001 (  3,  2) [002467] ----G--N----                 \--*  LCL_VAR   struct(AX) V107 tmp88        $455

CopyBlk based forward copy assertion for reverse copy assertion for [001471] V64 @00000003 by [001465] V64 @00000003.

***** (before)
N003 (  7,  5) [001468] -A--G---R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [001467] D---G--N----              +--*  LCL_VAR   struct(AX) V13 loc12        
N001 (  3,  2) [001465] ------------              \--*  LCL_VAR   struct V64 tmp45        u:2 (last use) $45c

N007 ( 13, 15) [001480] -A--G---R---              *  ASG       struct (copy) $VN.Void
N006 (  6,  7) [001479] n-----------              +--*  BLK(16)   struct
N005 (  3,  5) [001478] ------------              |  \--*  ADDR      byref  $c4d
N004 (  3,  4) [001471] D------N----              |     \--*  LCL_FLD   struct V64 tmp45        d:2[+0] Fseq[_value]
N003 (  6,  7) [001476] n---G-------              \--*  IND       struct $5c6
N002 (  3,  5) [001472] ----G-------                 \--*  ADDR      byref  $18f
N001 (  3,  4) [001475] -------N----                    \--*  LCL_FLD   struct V09 loc8         u:4[+0] Fseq[_value] (last use) $5c6

Copy propagated to:

***** (after)
N001 (  7,  5) [001468] ------------              *  NOP       void  

     ( 13, 15) [001481] ------------              *  STMT      void  (IL 0x18C...  ???)
N005 ( 10, 10) [001480] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N004 (  3,  2) [005876] D---G--N----                 +--*  LCL_VAR   struct(AX) V13 loc12        
N003 (  6,  7) [001476] n---G-------                 \--*  IND       struct $5c6
N002 (  3,  5) [001472] ----G-------                    \--*  ADDR      byref  $18f
N001 (  3,  4) [001475] -------N----                       \--*  LCL_FLD   struct V09 loc8         u:4[+0] Fseq[_value] (last use) $5c6

CopyBlk based forward copy assertion for reverse copy assertion for [001403] V09 @00000003 by [001475] V09 @00000003.

***** (before)
N005 ( 10, 10) [001480] -A--G---R---              *  ASG       struct (copy) $VN.Void
N004 (  3,  2) [005876] D---G--N----              +--*  LCL_VAR   struct(AX) V13 loc12        
N003 (  6,  7) [001476] n---G-------              \--*  IND       struct $5c6
N002 (  3,  5) [001472] ----G-------                 \--*  ADDR      byref  $18f
N001 (  3,  4) [001475] -------N----                    \--*  LCL_FLD   struct V09 loc8         u:4[+0] Fseq[_value] (last use) $5c6

N003 (  7,  5) [001404] -A------R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [001403] D------N----              +--*  LCL_VAR   struct V09 loc8         d:4
N001 (  3,  2) [001401] ------------              \--*  LCL_VAR   struct V60 tmp41        u:2 (last use) $58c

Copy propagated to:

***** (after)
N001 ( 10, 10) [001480] ------------              *  NOP       void  

     (  7,  5) [000562] ------------              *  STMT      void  (IL   ???...  ???)
N003 (  7,  5) [001404] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005877] D---G--N----                 +--*  LCL_VAR   struct(AX) V13 loc12        
N001 (  3,  2) [001401] ------------                 \--*  LCL_VAR   struct V60 tmp41        u:2 (last use) $58c

CopyBlk based forward copy assertion for reverse copy assertion for [001442] V60 @00000003 by [001401] V60 @00000003.

***** (before)
N003 (  7,  5) [001404] -A--G---R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005877] D---G--N----              +--*  LCL_VAR   struct(AX) V13 loc12        
N001 (  3,  2) [001401] ------------              \--*  LCL_VAR   struct V60 tmp41        u:2 (last use) $58c

N005 ( 10, 10) [001449] -A--G---R---              *  ASG       struct (copy) $VN.Void
N004 (  6,  7) [001448] n-----------              +--*  BLK(16)   struct
N003 (  3,  5) [001447] ------------              |  \--*  ADDR      byref  $c4c
N002 (  3,  4) [001442] D------N----              |     \--*  LCL_FLD   struct V60 tmp41        d:2[+0] Fseq[_value]
N001 (  3,  2) [001444] ----G--N----              \--*  LCL_VAR   struct(AX) V59 tmp40         $45b

Copy propagated to:

***** (after)
N001 (  7,  5) [001404] ------------              *  NOP       void  

     ( 10, 10) [001450] ------------              *  STMT      void  (IL 0x182...  ???)
N003 (  7,  5) [001449] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005878] D---G--N----                 +--*  LCL_VAR   struct(AX) V13 loc12        
N001 (  3,  2) [001444] ----G--N----                 \--*  LCL_VAR   struct(AX) V59 tmp40         $45b

CopyBlk based forward copy assertion for reverse copy assertion for [001056] V45 @00000003 by [001050] V45 @00000003.

***** (before)
N003 (  7,  5) [001053] -A--G---R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [001052] D---G--N----              +--*  LCL_VAR   struct(AX) V07 loc6         
N001 (  3,  2) [001050] ------------              \--*  LCL_VAR   struct V45 tmp26        u:2 (last use) $443

N007 ( 13, 15) [001065] -A--G---R---              *  ASG       struct (copy) $VN.Void
N006 (  6,  7) [001064] n-----------              +--*  BLK(16)   struct
N005 (  3,  5) [001063] ------------              |  \--*  ADDR      byref  $190
N004 (  3,  4) [001056] D------N----              |     \--*  LCL_FLD   struct V45 tmp26        d:2[+0] Fseq[_value]
N003 (  6,  7) [001061] n---G-------              \--*  IND       struct $5c0
N002 (  3,  5) [001057] ----G-------                 \--*  ADDR      byref  $18f
N001 (  3,  4) [001060] -------N----                    \--*  LCL_FLD   struct V09 loc8         u:2[+0] Fseq[_value] (last use) $5c0

Copy propagated to:

***** (after)
N001 (  7,  5) [001053] ------------              *  NOP       void  

     ( 13, 15) [001066] ------------              *  STMT      void  (IL 0x048...  ???)
N005 ( 10, 10) [001065] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N004 (  3,  2) [005879] D---G--N----                 +--*  LCL_VAR   struct(AX) V07 loc6         
N003 (  6,  7) [001061] n---G-------                 \--*  IND       struct $5c0
N002 (  3,  5) [001057] ----G-------                    \--*  ADDR      byref  $18f
N001 (  3,  4) [001060] -------N----                       \--*  LCL_FLD   struct V09 loc8         u:2[+0] Fseq[_value] (last use) $5c0

CopyBlk based forward copy assertion for reverse copy assertion for [000988] V09 @00000003 by [001060] V09 @00000003.

***** (before)
N005 ( 10, 10) [001065] -A--G---R---              *  ASG       struct (copy) $VN.Void
N004 (  3,  2) [005879] D---G--N----              +--*  LCL_VAR   struct(AX) V07 loc6         
N003 (  6,  7) [001061] n---G-------              \--*  IND       struct $5c0
N002 (  3,  5) [001057] ----G-------                 \--*  ADDR      byref  $18f
N001 (  3,  4) [001060] -------N----                    \--*  LCL_FLD   struct V09 loc8         u:2[+0] Fseq[_value] (last use) $5c0

N003 (  7,  5) [000989] -A------R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [000988] D------N----              +--*  LCL_VAR   struct V09 loc8         d:2
N001 (  3,  2) [000986] ------------              \--*  LCL_VAR   struct V41 tmp22        u:2 (last use) $581

Copy propagated to:

***** (after)
N001 ( 10, 10) [001065] ------------              *  NOP       void  

     (  7,  5) [000790] ------------              *  STMT      void  (IL   ???...  ???)
N003 (  7,  5) [000989] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005880] D---G--N----                 +--*  LCL_VAR   struct(AX) V07 loc6         
N001 (  3,  2) [000986] ------------                 \--*  LCL_VAR   struct V41 tmp22        u:2 (last use) $581

CopyBlk based forward copy assertion for reverse copy assertion for [001027] V41 @00000003 by [000986] V41 @00000003.

***** (before)
N003 (  7,  5) [000989] -A--G---R---              *  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005880] D---G--N----              +--*  LCL_VAR   struct(AX) V07 loc6         
N001 (  3,  2) [000986] ------------              \--*  LCL_VAR   struct V41 tmp22        u:2 (last use) $581

N005 ( 10, 10) [001034] -A--G---R---              *  ASG       struct (copy) $VN.Void
N004 (  6,  7) [001033] n-----------              +--*  BLK(16)   struct
N003 (  3,  5) [001032] ------------              |  \--*  ADDR      byref  $18e
N002 (  3,  4) [001027] D------N----              |     \--*  LCL_FLD   struct V41 tmp22        d:2[+0] Fseq[_value]
N001 (  3,  2) [001029] ----G--N----              \--*  LCL_VAR   struct(AX) V40 tmp21         $442

Copy propagated to:

***** (after)
N001 (  7,  5) [000989] ------------              *  NOP       void  

     ( 10, 10) [001035] ------------              *  STMT      void  (IL 0x03E...  ???)
N003 (  7,  5) [001034] -A--G---R---              \--*  ASG       struct (copy) $VN.Void
N002 (  3,  2) [005881] D---G--N----                 +--*  LCL_VAR   struct(AX) V07 loc6         
N001 (  3,  2) [001029] ----G--N----                 \--*  LCL_VAR   struct(AX) V40 tmp21         $442
```